src/lib/containers/provenance/ProvenanceGraph.tsx

```diff
@@ -0,0 +1,702 @@
+      nodesCopy: Node[]
+      edgesCopy: Edge[]
+    }) => {
+      const { entityHeaderIsCurrent, activity, nodesCopy, edgesCopy } = params
+      const entityNodeProps = {
+        type: NodeType.ENTITY,
+        data: entityHeaderIsCurrent,
+      }
+      if (isNodeNotFound(entityNodeProps, nodesCopy)) {
+        // add the new entity node
+        nodesCopy.push(getProvenanceNode(entityNodeProps))
+        if (activity) {
+          const activityNodeProps = {
+            type: NodeType.ACTIVITY,
+            data: activity,
+          }
+          edgesCopy.push(getProvenanceEdge(entityNodeProps, activityNodeProps))
+        }
+      }
+    },
+    [isNodeNotFound],
+  )
+
+  /**
+   * Called when we have a new entity to add to the graph.  This will result in adding
+   * a new Entity Node to the nodesCopy, a link from the usedInActivity to the new Entity Node,
+   * AND it will look for an Activity associated with the new entity.  If an Activity is found,
+   * it will add an Expand Node for the user to click (to expand to see this Activity!).
+   * If this is a root node (starting entity node), then we automatically add the Activity Node
+   * and process it.
+   */
+  const addEntity = useCallback(
+    async (params: {
+      entityHeaderIsCurrent: EntityHeaderIsCurrent
+      usedInActivity?: Activity
+      nodesCopy: Node[]
+      edgesCopy: Edge[]
+    }) => {
+      const { entityHeaderIsCurrent, nodesCopy, edgesCopy, usedInActivity } =
+        params
+      addEntityNode({
+        entityHeaderIsCurrent,
+        activity: usedInActivity,
+        nodesCopy,
+        edgesCopy,
+      })
+      //look for Activity
+      try {
+        const { entityHeader } = entityHeaderIsCurrent
+        const activity = await SynapseClient.getActivityForEntity(
+          entityHeader.id,
+          entityHeader.versionNumber,
+          accessToken,
+        )
+        // if this is not a root node (or there are too many items to show), add an expand node
+        if (
+          !isRootEntity(entityHeader) ||
+          (activity.used && activity.used.length >= MAX_ACTIVITY_EXPAND_NODES)
+        ) {
+          addExpandNode({
+            entityHeaderIsCurrent,
+            itemCount: activity.used?.length,
+            nodesCopy,
+            edgesCopy,
+          })
+        }
+      } catch (e) {
+        // Activity is not accessible
+        console.error(e)
+        if (isRootEntity(entityHeaderIsCurrent.entityHeader)) {
+          // add provenance undefined node
+          addUndefinedNode({
+            entityHeader: entityHeaderIsCurrent.entityHeader,
+            nodesCopy,
+            edgesCopy,
+          })
+        }
+      }
+    },
+    [addEntityNode, accessToken, isRootEntity, addExpandNode, addUndefinedNode],
+  )
+
+  /**
+   * This effect attempts to center the graph on one of the root nodes.
+   * Only initializes after after a root node has been added to the graph.
+   */
+  useEffect(() => {
```

Comment:
If any of these hooks are tightly coupled, you could consider moving them to one or more custom hooks, which could be tested atomically. It's not really clear to me how these hooks in particular might be better organized, though, so it might be only worth doing if you can think of a natural/clean way to do that.